### DISCUSSION TITLE
Is Servlet 4.0 JAR actually compatible with Jetty 9.4 in DW 2.x?

### DISCUSSION THREAD

**Discussion by robbytx** [dropwizard/dropwizard#5365] [General]
I've recently upgraded my application to DropWizard 2.1.0 from 1.3.29, and I've noticed that DropWizard now declares a [dependency on `jakarta.servlet-api:4.0.4`](https://github.com/dropwizard/dropwizard/blob/v2.1.0/dropwizard-dependencies/pom.xml#L42), which would imply it expects the web server to support the Servlet API 4.0 specification.

However, DropWizard currently ships with `jetty-server:9.4.46.v20220331`, which declares its own [dependency still on `javax.servlet:javax.servlet-api:3.1.0`](https://github.com/eclipse/jetty.project/blob/jetty-9.4.46.v20220331/pom.xml#L56), consistent with [its documentation](https://www.eclipse.org/jetty/documentation/jetty-9/index.html#what-jetty-version) showing that **Jetty 9.4 supports only the Servlet API 3.1 specification**.

It would appear that these two dependencies are in some sort of conflict although it is unclear the severity.  Clearly DropWizard 2.1, Jetty 9.4, and the Servlet API 4.0 JAR are able to function sufficiently well on the same classpath, or else lots of things would be broken.

However, as an example, I see that the `HttpServletRequest` interface in 4.0 declares a new method `getHttpServletMapping()` with a default implementation that returns default (null/empty) values for all its methods, with a comment in the Javadoc saying:
> The returned object is immutable. Servlet 4.0 compliant implementations must override this method.

Since `org.eclipse.jetty.server.Request` in Jetty 9.4 does not override this method, it would seem that a DropWizard 2.x application calling this method would discover the unexpected/incorrect values returned from the default implementations.  This is probably better than not failing outright, but subtle issues like this could definitely lead to confusion and application bugs.

I looked back in DropWizard history to find the point at which the Servlet API JAR was upgraded to 4.0 to see if there was some intentional thought put into the upgrade of the Servlet API JAR.  I found it simply occurred in https://github.com/dropwizard/dropwizard/pull/2395, during an upgrade to Jersey 2.27 without any commentary as to why.  Relatedly Jersey itself appears to have upgraded its own dependency to 4.0 in https://github.com/eclipse-ee4j/jersey/pull/4001, which went into the 2.28 release, similarly without any commentary as to why.  Both PRs seem to be of the general "let's upgrade our dependencies" variety, so perhaps neither project truly depends on any 4.0 APIs that require a Servlet 4.0 compliant server.  If there are any true dependencies on such APIs in particular scenarios, then those are likely to be broken in Jetty 9.4 / DropWizard 2.x.

TL;DR:

At this point, I suppose we're stuck with the status quo here until DropWizard can upgrade to Jetty 10/11, which actually support the Servlet API 4.0/5.0 versions.  In the meantime, DropWizard and application developers should take care when leveraging any APIs introduced in Servlet 4.0 to ensure they do not introduce bugs owing to Jetty 9's lack of support.

**Comment by zUniQueX** [comment]
Hi @robbytx. Thanks for reporting this!

The reason Jersey updated to `servlet-api:4.x` should be, that Jersey is being shipped with GlassFish. The [Servlet API 4.0 spec](https://jakarta.ee/specifications/servlet/4.0/) lists GlassFish 5.1 as a compatible implementation, so the shipped Jersey version (2.28 in this case) should work with `servlet-api:4.x`.

Dropwizard 2.1.x is a version we wanted to release before moving to Jetty 10/11, but we currently prepare these upgrades with Dropwizard 3.x and 4.x, as denoted in [this comment](https://github.com/dropwizard/dropwizard/discussions/4720#discussioncomment-2198975).

The first beta releases of Dropwizard 3.x and 4.x are already available and compatible with `servlet-api:4.x` and `servlet-api:5.x`.

**Comment by robbytx** [comment]
Thanks for your response.  I suppose it makes sense for Jersey 2.28 to use 4.x if that version is shipped with GlassFish 5.1.

My only concern here is whether Jersey 2.28-2.35 utilize any new Servlet 4.0 APIs that require a Servlet 4.0 compatible container.  I'm aware of (and love!) some of the ergonomic improvements like `HttpFilter` that work out of the box, but I'm mildly concerned there may be some latent issue in DropWizard 2.x due to its use of the Servlet 3.1-compliant Jetty 9.4.  

I suppose this is only a theoretical concern at this point, and as you said, the forthcoming versions of DropWizard should resolve this concern, so I'll just trust that there are no major issues in DW 2.x (or else I presume they would have been reported), and hope for the best!
